File: frontend/src/activity.js
// Activity log entries (GET /api/tasks/:id/activity) as sentences.
import { describeRecurrence } from './recurrence';

const STATUS = { todo: 'To Do', in_progress: 'In Progress', blocked: 'Blocked', done: 'Done' };
const PRIORITY = { 0: 'Low', 1: 'Medium', 2: 'High', 3: 'Critical' };

const status = (s) => STATUS[s] || s;
const priority = (p) => PRIORITY[p] || `P${p}`;

// "2026-10-01" -> local date text, without timezone shifts.
export function formatDay(iso) {
  const m = /^(\d{4})-(\d{2})-(\d{2})$/.exec(iso || '');
  return m ? new Date(+m[1], m[2] - 1, +m[3]).toLocaleDateString() : iso || '';
}

function repeatText(value) {
  const [unit, n] = (value || '').split(':');
  return describeRecurrence(unit, parseInt(n, 10) || 1).replace(/^Repeats /, '');
}

/** What happened, without the actor: "changed status from To Do to Done". */
export function describeActivity({ kind, actor, old_value: o, new_value: n }, fmtDay = formatDay) {
  switch (kind) {
    case 'created':
      return 'created the task';
    case 'title':
      return `renamed it from “${o}” to “${n}”`;
    case 'description':
      return 'edited the description';
    case 'status':
      if (n === 'done') return 'completed it';
      if (o === 'done') return `reopened it (${status(n)})`;
      return `changed status from ${status(o)} to ${status(n)}`;
    case 'priority':
      return `changed priority from ${priority(o)} to ${priority(n)}`;
    case 'deadline':
      if (!n) return 'removed the deadline';
      if (!o) return `set the deadline to ${fmtDay(n)}`;
      return `moved the deadline from ${fmtDay(o)} to ${fmtDay(n)}`;
    case 'assignee':
      if (!n) return o ? `unassigned ${o}` : 'unassigned it';
      if (n === actor) return o ? `took it over from ${o}` : 'took it on';
      if (!o) return `assigned it to ${n}`;
      return `reassigned it from ${o} to ${n}`;
    case 'project':
      return `moved it from ${o || 'No project'} to ${n || 'No project'}`;
    case 'labels':
      if (!o) return `labeled it ${n}`;
      if (!n) return `removed the labels (${o})`;
      return `changed labels from ${o} to ${n}`;
    case 'recurrence':
      return n ? `made it repeat ${repeatText(n)}` : 'stopped repeating it';
    case 'next_created':
      return n ? `created the next occurrence (due ${fmtDay(n)})` : 'created the next occurrence';
    case 'repeat_of':
      return 'created it as the next occurrence of a repeating task';
    default:
      return kind;
  }
}

File: backend/app/activity.py
"""Activity log per task: who created it and changed what, when. Values are
stored as display text at the time of the change (names, not ids), so the
history stays readable after users or projects are renamed or deleted."""
from sqlalchemy.orm import Session

from app.models import Project, Task, TaskActivity, User

# Changes worth showing; order, parent and completed_at are bookkeeping.
TRACKED = ("title", "description", "status", "priority", "deadline", "assignee_id", "project_id", "recurrence", "labels")


def _status(task: Task):
    return task.status.value if task.status is not None else None


def _recurrence(task: Task):
    return f"{task.recurrence_unit}:{task.recurrence_interval or 1}" if task.recurrence_unit else None


def snapshot(task: Task) -> dict:
    return {
        "title": task.title,
        "description": task.description or "",
        "status": _status(task),
        "priority": task.priority,
        "deadline": task.deadline.date().isoformat() if task.deadline else None,
        "assignee_id": task.assignee_id,
        "project_id": task.project_id,
        "recurrence": _recurrence(task),
        "labels": ", ".join(sorted(l.name for l in task.labels)) or None,
    }


def _user_name(db: Session, user_id):
    user = db.get(User, user_id) if user_id is not None else None
    return user.username if user else None


def _project_name(db: Session, project_id):
    project = db.get(Project, project_id) if project_id is not None else None
    if not project:
        return None
    parent = db.get(Project, project.parent_id) if project.parent_id else None
    return f"{parent.name} / {project.name}" if parent else project.name


def _text(value):
    if value is None:
        return None
    value = str(value)
    return value if len(value) <= 300 else value[:299] + "…"


def _add(db: Session, task: Task, actor: User, kind: str, old=None, new=None):
    db.add(TaskActivity(task_id=task.id, actor_id=actor.id if actor else None, kind=kind,
                        old_value=_text(old), new_value=_text(new)))


def created(db: Session, task: Task, actor: User):
    _add(db, task, actor, "created")


def changed(db: Session, task: Task, before: dict, actor: User):
    """One entry per tracked field that differs from `before` (a snapshot)."""
    after = snapshot(task)
    for field in TRACKED:
        old, new = before[field], after[field]
        if old == new:
            continue
        if field == "description":
            _add(db, task, actor, "description")  # the text itself would be noise
        elif field == "assignee_id":
            _add(db, task, actor, "assignee", _user_name(db, old), _user_name(db, new))
        elif field == "project_id":
            _add(db, task, actor, "project", _project_name(db, old), _project_name(db, new))
        else:
            _add(db, task, actor, field, old, new)


def repeated(db: Session, done: Task, nxt: Task, actor: User):
    """A repeating task was completed and created its next occurrence."""
    due = nxt.deadline.date().isoformat() if nxt.deadline else None
    _add(db, done, actor, "next_created", None, due)
    _add(db, nxt, actor, "repeat_of", done.id, due)
